lib/AST/Decl.cpp

```diff
@@ -1392,22 +1392,43 @@ static bool isPolymorphic(const AbstractStorageDecl *storage) {
   return false;
 }
 
+static bool isDirectToStorageAccess(const AccessorDecl *accessor,
+                                    const VarDecl *var, bool isAccessOnSelf) {
+  // All accesses have ordinary semantics except those to variables
+  // with storage from within their own accessors.
+  if (accessor->getStorage() != var)
+    return false;
+
+  if (!var->hasStorage())
+    return false;
+
+  // In Swift 5 and later, the access must also be a member access on 'self'.
+  if (!isAccessOnSelf &&
+      var->getDeclContext()->isTypeContext() &&
+      var->getASTContext().isSwiftVersionAtLeast(5))
+    return false;
+
+  // As a special case, 'modify' coroutines with forced static dispatch
+  // must use ordinary semantics, so that the 'modify' coroutine for a
+  // 'dynamic' property uses Objective-C message sends and not direct
+  // access to storage.
+  if (accessor->getAccessorKind() == AccessorKind::Modify &&
```

Comment:
Rather than checking for `modify`, this should probably include every accessor other than `get` or `set`. (It'd be a problem if we synthesized `read` as well.)

Reply:
Yeah. It would be enough to just check hasForcedStaticDispatch() - which also needs to be renamed. It means “I don’t have a vtable entry, I have shared linkage, and I dynamically dispatch getters and setters”. It’s only set for a read or modify on a dynamic property (for now at least). Got any suggestions for a better name?